test/Mono.Linker.Tests/TestCasesRunner/AssemblyChecker.cs

```diff
@@ -139,11 +142,82 @@ protected virtual void VerifyTypeDefinition (TypeDefinition original, TypeDefini
 			}
 		}
 
+		/// <summary>
+		/// Validates that all <see cref="KeptByAttribute"/> instances on a member are valid (i.e. the linker recorded a marked dependency described in the attribute)
+		/// </summary>
+		void VerifyKeptByAttributes (IMemberDefinition src, IMemberDefinition linked)
+		{
+			foreach (var keptByAttribute in src.CustomAttributes.Where (ca => ca.AttributeType.IsTypeOf<KeptByAttribute> ()))
+				VerifyKeptByAttribute (linked.FullName, keptByAttribute);
+		}
+
+		/// <summary>
+		/// Validates that all <see cref="KeptByAttribute"/> instances on an attribute provider are valid (i.e. the linker recorded a marked dependency described in the attribute)
+		/// <paramref name="src"/> is the attribute provider that may have a <see cref="KeptByAttribute"/>, and <paramref name="attributeProviderFullName"/> is the 'FullName' of <paramref name="src"/>.
+		/// </summary>
+		void VerifyKeptByAttributes (ICustomAttributeProvider src, string attributeProviderFullName)
+		{
+			foreach (var keptByAttribute in src.CustomAttributes.Where (ca => ca.AttributeType.IsTypeOf<KeptByAttribute> ()))
+				VerifyKeptByAttribute (attributeProviderFullName, keptByAttribute);
+		}
+
+		void VerifyKeptByAttribute (string keptAttributeProviderName, CustomAttribute attribute)
+		{
+			// public KeptByAttribute (string dependencyProvider, DependencyKind reason) { }
+			// public KeptByAttribute (Type dependencyProvider, DependencyKind reason) { }
+			// public KeptByAttribute (Type dependencyProvider, string memberName, DependencyKind reason) { }
+			Assert.AreEqual (nameof (KeptByAttribute), attribute.AttributeType.Name);
+			var expectedDependency = GetDependency ();
+			foreach (var dep in this.linkedTestCase.Customizations.DependencyRecorder.Dependencies) {
+				if (dep == expectedDependency) {
+					return;
+				}
+			}
+			string errorMessage = $"{keptAttributeProviderName} was expected to be kept by {expectedDependency.Source} with reason {expectedDependency.DependencyKind.ToString ()}.";
+			Assert.Fail (errorMessage);
+
+			TestDependencyRecorder.Dependency GetDependency ()
```

Comment:
Nit: Why make it into a local function - it's only used from one place and I think the code would read better if it was in-line.